apilib/service.py

```diff
@@ -101,23 +101,37 @@ def foo(self, foo_request):
             return FooResponse(...)
     '''
 
-    def invoke(self, method_name, request):
+    def invoke(self, method_name, request, already_validated=False):
         self.log_request(method_name, request)
 
         method_descriptor = self.resolve_method(method_name)
         method = getattr(self, method_descriptor.name)
-        try:
-            response = method(request)
-            response.response_code = ResponseCode.SUCCESS
-        except ApiException as e:
-            response = method_descriptor.response_class(response_code=e.response_code, errors=e.errors)
-        except AssertionError:
-            # Re-raise for assertions made in unittests
-            raise
-        except Exception as e:
-            if self.process_unhandled_exception(e):
+
+        if already_validated:
```

Comment:
Why not just skip this clause and put all the validation handling under `if not already_validated`?

Reply:
You mean like so?
```
        if not already_validated:
            error_context = validation.ErrorContext()
            validation_context = validation.ValidationContext(service=self.get_name(), method=method_name)
            request.validate(error_context, validation_context)
            validation_errors = error_context.all_errors()
            if validation_errors:
                response = method_descriptor.response_class(
                    response_code=ResponseCode.REQUEST_ERROR,
                    errors=[ApiError(code=ve.code, path=ve.path, message=ve.msg) for ve in validation_errors])
        else:
            try:
                response = method(request)
                response.response_code = ResponseCode.SUCCESS
            except ApiException as e:
                response = method_descriptor.response_class(response_code=e.response_code, errors=e.errors)
            except AssertionError:
                # Re-raise for assertions made in unittests
                raise
            except Exception as e:
                if self.process_unhandled_exception(e):
                    raise
                response = method_descriptor.response_class(response_code=ResponseCode.SERVER_ERROR)

        self.log_response(method_name, request, response)
        return response
```
It's kind of a tricky logic branch, but I think the problem there is that we need to actually run `method(request)` if there are no validation errors. Which we could do by removing that `else` above the try/except, but that's there because we don't return the response on a failed validation immediately, because we're always using the calls to `self.log_response` before returning at the bottom.

But might be worth just copying those two lines at the bottom to right after validation fails--definitely makes it a lot clearer, and doesn't duplicate much code.

Reply:
Hmm, I'm probably missing something silly. How come you can't just do the validation piece as a standalone thing? This basically:

```
        if do_validation:
            error_context = validation.ErrorContext()
            validation_context = validation.ValidationContext(service=self.get_name(), method=method_name)
            request.validate(error_context, validation_context)
            validation_errors = error_context.all_errors()
            if validation_errors:
                response = method_descriptor.response_class(
                    response_code=ResponseCode.REQUEST_ERROR,
                    errors=[ApiError(code=ve.code, path=ve.path, message=ve.msg) for ve in validation_errors])

        try:
            response = method(request)
            response.response_code = ResponseCode.SUCCESS
        except ApiException as e:
            response = method_descriptor.response_class(response_code=e.response_code, errors=e.errors)
        except AssertionError:
            # Re-raise for assertions made in unittests
            raise
        except Exception as e:
            if self.process_unhandled_exception(e):
                raise
            response = method_descriptor.response_class(response_code=ResponseCode.SERVER_ERROR)

        self.log_response(method_name, request, response)
        return response
```

Reply:
I think the problem there is that in the case where there are validation errors, we'll still try to run `method(request)`. Could resolve that by copying the two lines of log_response/return up to immediately below where response is first set, which I think may be the cleanest solution

Reply:
Oh, duh. Yes, agree with copying those two lines :)

Reply:
I thought it was throwing an exception there so it wouldn't matter.

Reply:
Sounds good--I'll try to do this and the rest of the tests tonight/tomorrow.